controllers/provider-alicloud/pkg/alicloud/client/types.go

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	alicloudvpc "github.com/aliyun/alibaba-cloud-sdk-go/services/vpc"
+)
+
+// Credentials are the credentials to access the Alicloud API.
+type Credentials struct {
+	AccessKeyID     string
+	AccessKeySecret string
+}
+
+// VPC is the interface to the Alicloud VPC service.
+type VPC interface {
+	// DescribeVPCs describes the VPCs for the request.
+	DescribeVPCs(req *alicloudvpc.DescribeVpcsRequest) (*alicloudvpc.DescribeVpcsResponse, error)
+	// DescribeNATGateways describes the NAT gateways for the request.
+	DescribeNATGateways(req *alicloudvpc.DescribeNatGatewaysRequest) (*alicloudvpc.DescribeNatGatewaysResponse, error)
+}
+
+// Factory is the factory to instantiate Alicloud clients.
+type Factory interface {
+	// NewVPCClientWithAccessKey creates a new VPC client from the given credentials and region.
+	NewVPCClientWithAccessKey(region, accessKeyID, accessKeySecret string) (VPC, error)
```

Comment:
That's not possible as far as I can see: If we have the factory interface like
```go
type Factory interface {
    NewVPCClientWithAccessKey(region, accessKeyID, accessKeySecret string) (VPC, error)
}
```
then this function is not compatible with that interface (go doesn't do automatic casts like that).

Reply:
The `Factory` interface should be changed as well, see above.

Reply:
I can't change the interface of the factory because then I'm not able to provide a mock implementation of the client if the factory always returns the concrete type.